GTMAppAuthSwift/Sources/AuthStateStore.swift

```diff
@@ -0,0 +1,150 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+
+/// Represents the operations needed to provide credential storage.
+@objc(GTMAuthStateStore)
+public protocol AuthStateStore {
+  /// The `String` name for the stored auth state.
+  @objc var itemName: String { get }
+
+  /// Creates an instance of an `AuthStateStore` conforming instance.
+  ///
+  /// - Parameter itemName: The `String` name for the auth state.
+  @objc init(itemName: String)
+
+  /// Saves the provided authorization.
+  ///
+  /// - Parameter authState: An instance of `AuthState` to save.
+  /// - Throws: Any error that may arise during the save.
+  /// - Note: This save operation will use the `credentialItemName` provided during initialization.
+  @objc func save(authState: AuthState) throws
+
+  /// Saves the provided authorization with the given `itemName`.
+  ///
+  /// - Parameter authState: An instance of `AuthState` to save.
+  /// - Throws: Any error that may arise during the save.
+  /// - Note: This save operation will _not_ use the `credentialItemName` provided during
+  ///     initialization.
+  @objc func save(authState: AuthState, forItemName itemName: String) throws
+
+  /// Removes the `AuthState` matching the provided `itemName`.
+  ///
+  /// - Parameter itemName: The `String` name to use while finding the `AuthState` to remove.
+  /// - Throws: Any error that may arise during the removal.
+  /// - Note: This removal will _not_ use the `credentialItemName` provided during initialization.
+  @objc func removeAuthState(withItemName itemName: String) throws
+
+  /// Removes the `AuthState`.
+  ///
+  /// - Throws: Any error that may arise during the removal.
+  /// - Note: This removal will use the `credentialItemName` provided during initialization.
+  @objc func removeAuthState() throws
+
+  /// Retrieves the `AuthState` matching the provided item name.
+  ///
+  /// - Parameter itemName: The `String` name for the `authState` to retrieve from the store.
+  /// - Throws: Any error that may arise during the retrieval.
+  /// - Note: This retrieval will _not_ use the `credentialItemName` provided during initialization.
+  @objc func authState(forItemName itemName: String) throws -> AuthState
+
+  /// Retrieves the `authState`.
+  ///
+  /// - Throws: Any error that may arise during the retrieval.
+  /// - Note: This retrieval will use the `credentialItemName` provided during initialization.
+  @objc func retrieveAuthState() throws -> AuthState
+}
+
+/// Represents the operations needed to convert OAuth2 authorizations to `AuthState` instances along
+/// with credential storage.
+@objc(GTMOAuth2AuthStateStore)
+public protocol OAuth2AuthStateStore {
```

Comment:
I still feel like there's an opportunity to bring `OAuth2AuthStateCompatibility` more in line with `KeychainStore` by having them implement the same protocol.  `OAuth2AuthStateCompatibility` could expose some additional helper methods to cover the GTMOAuth2 persistence string methods, otherwise it seems to line up pretty well with `AuthStateStore`.

Reply:
Disregard my previous comment, I didn't notice that `KeychainStore` implements this protocol in addition to `AuthStateStore`.

A few thoughts...

- Use consistent method names: `[retrieve | save | remove]GTMOAuth2Format`?
- We should refer to the old SDK as 'GTMOAuth2' rather than `OAuth2`.
  - Might want to make the name of this class `GTMOAuth2AuthStateStore` in Swift as well as Objc.

Reply:
Name changes made.